test/module/libs/common/combine_latest_until_first_completed_test.cpp

```diff
@@ -0,0 +1,237 @@
+    size_t current_observable_idx,
+    bool &completed,
+    long current_test_value,
+    Types... args) {
+  if (current_observable_idx == 0) {
+    EXPECT_FALSE(completed) << "Got values when already completed!";
+  }
+  checkValues(observable_helpers,
+              current_observable_idx,
+              completed,
+              current_test_value);
+  checkValues(
+      observable_helpers, current_observable_idx + 1, completed, args...);
+}
+
+template <>
+void checkValues(
+    const std::vector<IntervalObservebleHelperVariant> &observable_helpers,
+    size_t current_observable_idx,
+    bool &completed,
+    long current_test_value) {
+  auto last_value =
+      iroha::visit_in_place(observable_helpers[current_observable_idx],
+                            [](const auto &observable_helper) {
+                              return observable_helper.lastValue();
+                            });
+  EXPECT_EQ(last_value, current_test_value)
+      << "Got wrong value from observable #" << current_observable_idx;
+
+  iroha::visit_in_place(
+      observable_helpers[current_observable_idx],
+      [=, &completed](const IntervalObservebleTakeNHelper &observable_helper) {
+        const long completion_value = observable_helper.taken_;
+        EXPECT_LE(last_value, completion_value)
+            << "Got value after the last one from observable #"
+            << current_observable_idx;
+        completed |= completion_value <= last_value;
+      },
+      [](const auto &) {});
+}
+
+auto makeValuesChecker(
+    const std::vector<IntervalObservebleHelperVariant> &observable_helpers) {
+  return [&observable_helpers](auto... values) {
+    bool completed;
+    checkValues(observable_helpers, 0, completed, values...);
+    return completed;
+  };
+}
+
+template <class T>
+rxcpp::observable<long> getObservable(T &observable_helper) {
+  return iroha::visit_in_place(observable_helper, [](auto &observable_helper) {
+    return observable_helper.getObservable();
+  });
+}
+
+/**
+ * Checks that correct values are emitted and that observing stops as soon as
+ * any of them completes.
+ * @return whether the final value of any source observable was reached
+ */
+bool check(rxcpp::observable<bool> observable) {
+  bool completed;
+  observable.last().as_blocking().subscribe(
+      [&completed](bool val) { completed = val; },
+      [](std::exception_ptr ep) {
+        try {
+          std::rethrow_exception(ep);
+        } catch (const std::exception &e) {
+          FAIL() << "OnError: " << e.what();
+        }
+      });
+  return completed;
+}
+
+/**
+ * @given three observables
+ * @when they are combined in parallel
+ * @then the returned values are correct
+ * @and the pipeline stops as soon as the first observable stops.
+ */
+TEST(combineLatestUntilFirstCompleted, ThreeParallel) {
+  std::vector<IntervalObservebleHelperVariant> observable_helpers{
+      IntervalObservebleHelper(17ms),
```

Comment:
The test cannot be built on mac environment.

Please check the log https://gist.github.com/igor-egorov/9fde3d271c09975db87163f44ea70b92